### ISSUE TITLE
kubelet: NodePublishVolumeRequest.target_path handling not CSI compliant

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#75535]
**What happened**:

When mounting a filesystem volume with a CSI driver, kubelet creates the `NodePublishVolumeRequest.target_path` directory where the volume is to be mounted.

**What you expected to happen**:

According to the CSI 1.0 spec, kubelet must *not* create that directory, only its parent:
```
The CO SHALL ensure that the parent directory of this path exists
and that the process serving the request has `read` and `write`
permissions to that parent directory.
For volumes with an access type of block, the SP SHALL place the
block device at target_path.
For volumes with an access type of mount, the SP SHALL place the
mounted directory at target_path.
```

**Environment**:
- Kubernetes version (use `kubectl version`): Kubernetes 1.13.x and 1.14.0

/sig storage
/cc @msau42 @bswartz @lpabon @vladimirvivien

**Comment by pohly** [comment]
Related fix in csi-test: https://github.com/kubernetes-csi/csi-test/pull/182

**Comment by msau42** [comment]
I'm not sure how we can fix this without potentially breaking drivers that were dependent on the old mechanism.  Maybe we can put out a deprecation notice and then fix it after the deprecation period?

**Comment by pohly** [comment]
@msau42 that sounds like a reasonable approach.

**Comment by bswartz** [comment]
My proposal to fix this is to relax the spec. Kubernetes already delete the directory if the SP violates the spec and leave the directory behind after nodeunpublish. Many (all?) SPs already check if the directory exists before creating it during nodepublish.

Thus we can make 2 relaxing changes to the CSI spec and escape from this problem without changing Kubernetes behavior at all:
1. Change the spec wording to say that the CO MAY create the directory before calling NodePublish. This would in theory cause breakage for plugins that rely on the CO not creating the directory (the plugin would have to be exceptionally strict) but since this is the existing behavior most plugins would accept this.
2. Change the wording in the spec to say that the SP SHOULD (rather than MUST) delete the directory on unpublish. This could in theory break COs that rely on the SP to delete the directory, but I suspect that some SPs already do not delete the directory, forcing CO's to deal with this noncompliant behavior anyways.

**Comment by pohly** [comment]
I agree with @bswartz. The drivers I work on now implement exactly this proposal, including the deletion during `NodeUnpublishVolume` and that works fine with Kubernetes.

**Comment by bswartz** [comment]
I plan to bring this up at the May 29 meeting.

**Comment by sftim** [comment]
What priority should this issue have?

**Comment by msau42** [comment]
/lifecycle frozen
We plan to fix this with https://github.com/kubernetes/kubernetes/pull/88759

**Comment by msau42** [comment]
/close

Fix was made in #88759 in 1.20 with deprecation notice sent out in 1.18.
